src/Voron/Data/CompactTrees/PersistentDictionary.cs

```diff
@@ -0,0 +1,199 @@
+            where TKeysEnumerator : struct, IReadOnlySpanEnumerator
+        {
+            // Reserve the memory to create the encoder table. 
+            using var scope = llt.Allocator.Allocate(UsableDictionarySize, out var output);
+            
+            // Train the new dictionary.
+            var encoder = new HopeEncoder<Encoder3Gram<NativeMemoryEncoderState>>(
+                                new Encoder3Gram<NativeMemoryEncoderState>(
+                                    new NativeMemoryEncoderState(output.Ptr, UsableDictionarySize)));
+            encoder.Train(trainEnumerator, MaxDictionarySize);
+
+            // Test the new dictionary to ensure that we have statistically better compression.
+            using var encodeBufferScope = llt.Allocator.Allocate(Constants.Storage.PageSize, out var encodeBuffer);
+            
+            int incumbentSize = 0;
+            int successorSize = 0;
+            var auxEncodeBuffer = encodeBuffer.ToSpan();
+            while (testEnumerator.MoveNext(out var testValue))
+            {                
+                incumbentSize += previousDictionary._encoder.Encode(testValue, auxEncodeBuffer);
+                successorSize += encoder.Encode(testValue, auxEncodeBuffer);
+            }
+
+            // If the new dictionary is not at least 2.5% better, we return the current dictionary.             
+            if (incumbentSize < successorSize * 1.025)
+                return previousDictionary;            
+                                    
+            var p = llt.AllocatePage(NumberOfPagesForDictionary);
+            p.Flags = PageFlags.Overflow;
+            p.OverflowSize = NumberOfPagesForDictionary * Constants.Storage.PageSize;
+
+            PersistentDictionaryHeader* header = (PersistentDictionaryHeader*)p.DataPointer;
+            header->CurrentId = p.PageNumber;
+            header->PreviousId = previousDictionary != null ? previousDictionary.PageNumber : 0;
+            output.CopyTo(0, p.DataPointer + PersistentDictionaryHeader.SizeOf, 0, UsableDictionarySize);
+            header->TableHash = XXHash64.Calculate(p.DataPointer + sizeof(long), UsableDictionarySize + PersistentDictionaryHeader.SizeOf - sizeof(long));
+
+#if DEBUG
+            VerifyTable(p);
+#endif
+
+            return new PersistentDictionary(p);
+        }
+
+        public static void VerifyTable(Page page)
+        {
+            var tableHash = XXHash64.Calculate(page.DataPointer + sizeof(long), UsableDictionarySize + PersistentDictionaryHeader.SizeOf - sizeof(long));
+            PersistentDictionaryHeader* header = (PersistentDictionaryHeader*)page.DataPointer;
+
+            // If checksum is not correct, we will throw a corruption exception
+            if (tableHash != header->TableHash)
+                throw new VoronErrorException($"Persistent storage checksum mismatch, expected: {tableHash}, actual: {header->TableHash}");
```

Comment:
What could actually cause that? You aren't really verifying the data, only that this is matching to the hash.
We already do that at the page level

Reply:
This is a debug/verification check only, I do a lot of memory copying and handling. When modifying this I need to ensure certain things match. Runtime impact of having the hash is negligible in both space and time.